### PR TITLE
Allow disabling of ring heartbeats by setting relevant options to zero.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
Allow disabling of ring heartbeats by setting relevant options to zero.

Depends on https://github.com/cortexproject/cortex/pull/4342

**Checklist**
- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,13 @@
 
 * [CHANGE] Querier / ruler: Change `-querier.max-fetched-chunks-per-query` configuration to limit to maximum number of chunks that can be fetched in a single query. The number of chunks fetched by ingesters AND long-term storare combined should not exceed the value configured on `-querier.max-fetched-chunks-per-query`. #4260
 * [ENHANCEMENT] Add timeout for waiting on compactor to become ACTIVE in the ring. #4262
+* [ENHANCEMENT] Ring: allow heartbeats to be explicitly disabled by setting the interval to zero. This is considered experimental. This applies to the following configuration options: #4344
```
Comment (pracucci) [inline CHANGELOG.md:7]: Mention it's experimental in the v1 guarantees doc too, please.

**Review by pracucci — COMMENTED** [review]
The solution you've picked should work.

Initially, I thought that we should not set Timestamp when the instance is registered in the ring and we should not update it when an InstanceDesc is updated, but then I've realised memberlist actually requires that Timestamp change each time we do update InstanceDesc so we have to keep all that logic and just disabled the hearbeating part.

However, I would update the ring web page to clearly state when heartbeating is disabled.

**Review by tomwilkie — APPROVED** [review]
LGTM;

+1 to Marco's suggestion on including the heartbeat value / disabled on the ring info page.

Not related to this change, but why is there so much duplication of ring config now?

**Comment by stevesg** [comment]
> Initially, I thought that we should not set Timestamp when the instance is registered in the ring and we should not update it when an InstanceDesc is updated, but then I've realised memberlist actually requires that Timestamp change each time we do update InstanceDesc so we have to keep all that logic and just disabled the hearbeating part.

I had the same worry that memberlist relies on the Timestamp to resolve conflicts but it looks that the Timestamp is updated whenever the ring state is updated ([example](https://github.com/cortexproject/cortex/blob/f8b08a367dd0e4aeb0ce1fb3e59544ad82a7ccaa/pkg/ring/lifecycler.go#L339)). This makes sense as in my testing with very long heartbeats, I didn't encounter any issues.

Will update the status page

**Comment by stevesg** [comment]
It doesn't seem trivial to add the heartbeat period to the status page, as the configuration actually exists in the lifecycler :/ Any suggestions?

**Comment by pracucci** [comment]
> It doesn't seem trivial to add the heartbeat period to the status page, as the configuration actually exists in the lifecycler :/ Any suggestions?

The easiest (but probably) hackiest one is to add the heartbeat period to `ring.Config` too, but not registered via a CLI flag and not exposed via YAML. Then, in the code, we do set it based on the setting coming from the lifecycler.

**Review by bboreham — APPROVED** [review]

**Comment by bboreham** [comment]
> Not related to this change, but why is there so much duplication of ring config now?

Can't speak to the history but I agree there is a lot of duplication?
Maybe because the first variant didn't want the "lifecycler" behaviour?

**Comment by pracucci** [comment]
> > Not related to this change, but why is there so much duplication of ring config now?
> 
> Can't speak to the history but I agree there is a lot of duplication?
> Maybe because the first variant didn't want the "lifecycler" behaviour?

I'm the guilty one. Long time ago (when we built the blocks storage) we duplicated (once) the ring config to slim it down. The `ring.Config` has a bunch of options we didn't need for the compactor so to make the config more clear (instead of reusing the same config and having many options not working for compactor) we duplicated it. The same pattern than has been applied for other rings. It's something we can definitely cleanup.

**Review by pracucci — APPROVED** [review]
The improvement to the HTTP ring page can be done in a follow up PR.

**Comment by edma2** [comment]
Out of curiosity, what is the motivation behind this PR (and https://github.com/cortexproject/cortex/pull/4342)? I couldn't find much docs on this feature. Does this improve availability by ignoring heartbeat failures?

**Comment by bboreham** [comment]
Updating a KV store every few seconds is difficult to scale past 100-200 ingesters. 
As of #741 we have a heartbeat directly from distributors to ingesters, so the one via the ring should not be necessary.

Some of the story is here: https://grafana.com/blog/2020/02/11/how-were-abusing-hashicorps-consul-at-grafana-labs/

**Pull request merged**